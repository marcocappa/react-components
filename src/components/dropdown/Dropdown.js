import { useState, useEffect, useRef } from 'react';
import classnames from 'classnames';
import IconArrows from '../icons/IconArrows';
import IconClose from '../icons/IconClose';
import './Dropdown.scss';

const Dropdown = ({ id: selectId = '', options, label, onSelect }) => {
	const dropdownRef = useRef();
	const dropdownButtonRef = useRef();
	const [filteredOptions, setFilteredOptions] = useState([]);
	const [valueSelected, setValueSelected] = useState('');
	const [idSelected, setIdSelected] = useState('');
	const [showOptions, setShowOptions] = useState(false);
	const [searchedOption, setSearchedOption] = useState('');

	const handleShowOptions = ({ target }) => {
		// if click or focus on the button showOptions
		if (dropdownRef.current && dropdownRef.current.contains(target)) {
			if (target === dropdownButtonRef.current) {
				setShowOptions((s) => !s);
			} else {
				setShowOptions(true);
			}
			return;
		}
		// if click or focus outside the "ref" hideOptions
		setShowOptions(false);
	};

	const handleFocus = () => {
		setShowOptions(true);
	};

	const handleSelection = (value, id) => {
		// set value selected and id:
		setValueSelected(value);
		setIdSelected(id);
		// close dropdown list:
		setShowOptions(false);
		// call onSelect with {id and value}:
		onSelect({ id, value });
	};

	const handleFocusOut = ({ target }) => {
		if (dropdownRef.current && dropdownRef.current.contains(target)) {
			return;
		}
		setShowOptions(false);
	};

	const handleCloseOptions = () => {
		setShowOptions(false);
	};

	const handleFilterOptions = ({ target: { value } }) => {
		setSearchedOption(value);
	};

	const handleResetSerchedOption = () => setSearchedOption('');

	useEffect(() => {
		document.addEventListener('keyup', handleFocusOut);
		document.addEventListener('mousedown', handleShowOptions);
		window.addEventListener('blur', handleCloseOptions);
		return () => {
			document.removeEventListener('keyup', handleFocusOut);
			document.removeEventListener('mousedown', handleShowOptions);
			window.removeEventListener('blur', handleCloseOptions);
		};
	}, []);

	useEffect(() => {
		setFilteredOptions(options);
	}, [options]);

	useEffect(() => {
		const newOptions = [...options].filter((option) =>
			option.value.toLowerCase().includes(searchedOption.toLowerCase())
		);
		console.log(newOptions);
		setFilteredOptions(newOptions);
	}, [options, searchedOption]);

	return (
		<div className="dropdown" ref={dropdownRef}>
			<div className="dropdown__header">
				{label && <label className="dropdown__header-label">{label}</label>}
				<button
					ref={dropdownButtonRef}
					className="dropdown__header-selected-option"
					onFocus={handleFocus}
				>
					{valueSelected === '' ? 'Select' : valueSelected}
					<IconArrows className="dropdown__header-icon" />
				</button>
			</div>
			<div className={classnames('dropdown__list', showOptions && 'active')}>
				{options.length > 10 && (
					<div className="dropdown__list-wrap-search">
						<input
							name={`select-option-search-${selectId}`}
							className="dropdown__list-search"
							type="text"
							placeholder="Search..."
							autoComplete="off"
							value={searchedOption}
							onChange={handleFilterOptions}
						/>

						<button
							className={classnames(
								'dropdown__list-clear-search',
								searchedOption && 'active'
							)}
							onClick={handleResetSerchedOption}
						>
							<IconClose className="dropdown__list-clear-search-icon" />
						</button>
					</div>
				)}
				{filteredOptions.length > 0 ? (
					filteredOptions.map(({ value, id }) => (
						<button
							className={classnames(
								'dropdown__list-item',
								idSelected === id && 'selected'
							)}
							key={id}
							onClick={() => handleSelection(value, id)}
						>
							{value}
						</button>
					))
				) : (
					<p className="dropdown__list-no-options">No options available!</p>
				)}
			</div>
		</div>
	);
};

export default Dropdown;
