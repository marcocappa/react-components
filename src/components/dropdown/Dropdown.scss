.dropdown {
    background-color: transparent;
    position: relative;

    &__header {
        &-label {
            margin: 4px 4px 10px;
            color: #999999;
            text-transform: uppercase;
            display: inline-block;
            font-weight: 500;
        }
        &-selected-option {
            cursor: pointer;
            border-radius: 4px;
            color: black;
            width: 100%;
            box-sizing: border-box;
            padding: 10px 12px;
            background-color: white;
            border: 2px solid rgb(207, 220, 238);
            margin-bottom: 4px;
            text-align: left;
            font-weight: 300;
            position: relative;
            box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
        }
        &-icon {
            position: absolute;
            right: 10px;
        }
    }

    &__list {
        background-color: white;
        border: 1px solid rgb(207, 220, 238);
        max-height: 390px;
        overflow-y: scroll;
        display: none;
        box-sizing: border-box;
        position: absolute;
        z-index: 1;
        width: 100%;
        box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;

        &-wrap-search {
            width: 100%;
            box-sizing: border-box;
            padding: 0.75rem;
            position: relative;
        }

        &-search {
            cursor: pointer;
            border-radius: 4px;
            color: black;
            width: 100%;
            box-sizing: border-box;
            padding: 10px 12px;
            border: 2px solid rgb(207, 220, 238);
            margin-bottom: 4px;
            font-weight: 300;
        }

        &-clear-search {
            width: 20px;
            height: 20px;
            position: absolute;
            right: 20px;
            top: 22px;
            border: none;
            display: none;
            background-color: transparent;
            justify-content: center;
            align-items: center;
            cursor: pointer;

            &.active {
                display: flex;
            }
        }

        &-clear-search-icon {
            width: 10px;
            height: 10px;
        }

        &-item {
            cursor: pointer;
            width: 100%;
            border: none;
            border-top: 1px solid rgb(207, 220, 238);
            background-color: transparent;
            padding: 10px 12px;
            text-align: left;
            font-weight: 300;

            &:hover {
                background-color: #f4f4f4;
            }

            &.selected {
                background-color: #f4f4f4;
            }
        }

        &-no-options {
            margin: 0;
            padding: 0 12px 12px;
            font-weight: 300;
        }

        &.active {
            display: block;
        }
    }
}
