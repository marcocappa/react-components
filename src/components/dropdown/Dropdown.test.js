import { render } from '@testing-library/react';
import Dropdown from './Dropdown';

test('renders dropdown', () => {
	const handleUserSelect = jest.fn();
	const options = [
		{ value: 'Andy', id: 1 },
		{ value: 'George', id: 2 },
		{ value: 'Clive', id: 3 },
		{ value: 'Peter', id: 4 },
		{ value: 'Cleatus', id: 5 },
		{ value: 'Alex', id: 6 },
		{ value: 'Omar', id: 7 },
		{ value: 'Carl', id: 8 },
		{ value: 'William', id: 9 },
		{ value: 'Donald', id: 10 },
		{ value: 'Dave', id: 11 },
		{ value: 'Jerry', id: 12 },
		{ value: 'Gerrie', id: 13 },
	];

	const { asFragment } = render(
		<Dropdown
			id="users"
			options={options}
			onSelect={handleUserSelect}
			label="Select user:"
		/>
	);

	expect(asFragment()).toMatchSnapshot();
});
