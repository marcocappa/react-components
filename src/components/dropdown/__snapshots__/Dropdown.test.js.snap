// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders dropdown 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <div
      class="dropdown__header"
    >
      <label
        class="dropdown__header-label"
      >
        Select user:
      </label>
      <button
        class="dropdown__header-selected-option"
      >
        Select
        <svg
          class="dropdown__header-icon"
          height="16px"
          width="8px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="none"
            stroke-width="1"
          >
            <g
              class="citi-table-sortable__icon"
              fill="#BFBFBF"
              fill-rule="nonzero"
            >
              <polygon
                class="arrow-desc"
                points="8 16 4 10 -2.27373675e-13 16"
                transform="translate(4.000000, 13.000000) scale(1, -1) translate(-4.000000, -13.000000) "
              />
              <polygon
                class="arrow-asc"
                points="8 6 4 -1.59872116e-14 -2.27373675e-13 6"
              />
            </g>
          </g>
        </svg>
      </button>
      <button
        class="dropdown__header-clear-select"
      >
        <svg
          class="dropdown__header-clear-select-icon"
          height="329pt"
          viewBox="0 0 329.26933 329"
          width="329pt"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m194.800781 164.769531 128.210938-128.214843c8.34375-8.339844 8.34375-21.824219 0-30.164063-8.339844-8.339844-21.824219-8.339844-30.164063 0l-128.214844 128.214844-128.210937-128.214844c-8.34375-8.339844-21.824219-8.339844-30.164063 0-8.34375 8.339844-8.34375 21.824219 0 30.164063l128.210938 128.214843-128.210938 128.214844c-8.34375 8.339844-8.34375 21.824219 0 30.164063 4.15625 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921875-2.089844 15.082031-6.25l128.210937-128.214844 128.214844 128.214844c4.160156 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921874-2.089844 15.082031-6.25 8.34375-8.339844 8.34375-21.824219 0-30.164063zm0 0"
          />
        </svg>
      </button>
    </div>
    <div
      class="dropdown__list"
    >
      <div
        class="dropdown__list-wrap-search"
      >
        <input
          autocomplete="off"
          class="dropdown__list-search"
          name="select-option-search-users"
          placeholder="Search..."
          type="text"
          value=""
        />
        <button
          class="dropdown__list-clear-search"
        >
          <svg
            class="dropdown__list-clear-search-icon"
            height="329pt"
            viewBox="0 0 329.26933 329"
            width="329pt"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m194.800781 164.769531 128.210938-128.214843c8.34375-8.339844 8.34375-21.824219 0-30.164063-8.339844-8.339844-21.824219-8.339844-30.164063 0l-128.214844 128.214844-128.210937-128.214844c-8.34375-8.339844-21.824219-8.339844-30.164063 0-8.34375 8.339844-8.34375 21.824219 0 30.164063l128.210938 128.214843-128.210938 128.214844c-8.34375 8.339844-8.34375 21.824219 0 30.164063 4.15625 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921875-2.089844 15.082031-6.25l128.210937-128.214844 128.214844 128.214844c4.160156 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921874-2.089844 15.082031-6.25 8.34375-8.339844 8.34375-21.824219 0-30.164063zm0 0"
            />
          </svg>
        </button>
      </div>
      <button
        class="dropdown__list-item"
      >
        Andy
      </button>
      <button
        class="dropdown__list-item"
      >
        George
      </button>
      <button
        class="dropdown__list-item"
      >
        Clive
      </button>
      <button
        class="dropdown__list-item"
      >
        Peter
      </button>
      <button
        class="dropdown__list-item"
      >
        Cleatus
      </button>
      <button
        class="dropdown__list-item"
      >
        Alex
      </button>
      <button
        class="dropdown__list-item"
      >
        Omar
      </button>
      <button
        class="dropdown__list-item"
      >
        Carl
      </button>
      <button
        class="dropdown__list-item"
      >
        William
      </button>
      <button
        class="dropdown__list-item"
      >
        Donald
      </button>
      <button
        class="dropdown__list-item"
      >
        Dave
      </button>
      <button
        class="dropdown__list-item"
      >
        Jerry
      </button>
      <button
        class="dropdown__list-item"
      >
        Gerrie
      </button>
    </div>
  </div>
</DocumentFragment>
`;
